Configuring middleware in the component (#3937)
dapr/docs: daprdocs/content/en/operations/components/middleware.md

title: "Configure middleware components"
linkTitle: "Configure middleware"
weight: 2000
description: "Customize processing pipelines via the middleware components"
---

{{% alert title="Note" color="primary" %}}
Configuring middleware components using the [configuration schema]({{< ref configuration-schema.md >}}) is **deprecated**.
{{% /alert %}}

> With Dapr, you can define custom processing pipelines by chaining a series of middleware components. There are two places that you can use a middleware pipeline:

Reviewer: ```suggestion
With Dapr, you can define custom processing pipelines by chaining a series of middleware components. There are two places that you can use a middleware pipeline:
```

we are not adding the middleware in series to follow the pipeline .

1. Building block APIs - HTTP middleware components are executed when invoking any Dapr HTTP APIs.

> 1. Service-to-Service invocation - HTTP middleware components are applied to service-to-service invocation calls.

Reviewer: ```suggestion
1. Service-to-Service invocation - appHttp middleware components are applied to service-to-service invocation calls.
```

> In this guide, you learn how to configure middleware using its component YAML. To learn how to create middleware components, see the [Author middleware components how-to guides.]({{< ref develop-middleware.md >}}) 

Reviewer: ```suggestion
In this guide, you learn how to configure middleware component . To learn how to create middleware components, see the [Author middleware components how-to guides.]({{< ref develop-middleware.md >}}) 
```

> title: "Configure middleware components"

Reviewer 2: Suggest just calling this Middleware components now, since these are just set in the normal component approach going forward. See comments below.
```suggestion
title: "Middleware components"
```

> linkTitle: "Configure middleware"

Reviewer 2: ```suggestion
linkTitle: "Middleware"
```

weight: 2000

> description: "Customize processing pipelines via the middleware components"

Reviewer 2: ```suggestion
description: "Customize processing pipelines using middleware components"
```

---

{{% alert title="Note" color="primary" %}}

> Configuring middleware components using the [configuration schema]({{< ref configuration-schema.md >}}) is **deprecated**.

Reviewer 2: ```suggestion
Configuring middleware components using a [configuration file]({{< ref configuration-schema.md >}}) is **no longer recommended**. Middleware components are treated same as other [components]({{< ref components-concept.md >}}) with a yaml file that is placed into the application resources folder.
```

Reviewer 2: All the rest of the changes in this file can be reverted.  In Summary we need to 
1) Say there is two pipelines -  httpPipeline and appHttpPipeline and move the existing pipeline diagrams to this section
2) Say there are two way to configure middleware. An method using configuration files which is not longer recommended or just as a Dapr component, which is the recommended way.
3) Have a new section which describes how to set a middleware component (i.e. includes the priority and pipelinetype field) similar to these sections https://docs.dapr.io/operations/components/setup-bindings/
4) Keep the "Configure API middleware pipelines" 
 and "Configure app middleware pipelines " https://docs.dapr.io/operations/components/middleware/#configure-api-middleware-pipelines" headings but make these sub headings under a new heading called " Using middleware components with configuration"
5) Check through all the other docs for updates on middleware. For example
 - Update this diagram to replace observability with middleware https://docs.dapr.io/concepts/components-concept/#component-specification
 - Update the https://docs.dapr.io/reference/components-reference/supported-middleware/ table to add a column "since runtime version"

<img src="/images/middleware.png" width="800" alt="Diagram showing the flow of a request and a response through the middlewares, as described in the paragraph above" />

HTTP middleware components are executed when invoking Dapr HTTP APIs using the `httpPipeline` setting.

> The following example defines a custom pipeline that uses a [RouterChecker middleware]({{< ref middleware-routerchecker.md >}}). In this case, all requests are authorized against the `^[A-Za-z0-9/._-]+$`:regex before they are forwarded to user code. The `priority` field determines the order in which requests are executed once all handler components are collected.

Reviewer: ```suggestion
The following example defines a custom pipeline that uses a [RouterChecker middleware]({{< ref middleware-routerchecker.md >}}). In this case, all requests are authorized to follow the regex rule `^[A-Za-z0-9/._-]+$` before they are forwarded to user code. The `priority` field determines the order in which requests are executed once all handler components are collected.
```
Also in brackets we can put "make sure to follow different priority for different middleware components else dapr might set it randomly"

<img src="/images/app-middleware.png" width="800" alt="Diagram showing the flow of a service invocation request. Requests from the callee Dapr sidecar to the callee application go through the app middleware pipeline as described in the paragraph above." />

> Any middleware component that can be used as HTTP middleware can also be applied to service-to-service invocation calls as a middleware component using the `appHttpPipeline` configuration. The example below adds the `routerchecker` middleware component for all outgoing calls from the Dapr sidecar (target of service invocation) to the application that this configuration is applied to.

Reviewer 2: revert this and keep the existing
